From 2210cf79c6f4df9a964a634821bcc456d6632e54 Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Mon, 28 Apr 2014 18:22:39 +0800
Subject: [PATCH 22/43] ENGR00307992-2 use KHR fence when current fence is
 invalid

this Android native issue caused abnormal display in large font

Signed-off-by: Xianzhong <b07117@freescale.com>
---
 libs/gui/GLConsumer.cpp | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/libs/gui/GLConsumer.cpp b/libs/gui/GLConsumer.cpp
index 3d9b6c4..d819a79 100644
--- a/libs/gui/GLConsumer.cpp
+++ b/libs/gui/GLConsumer.cpp
@@ -1014,6 +1014,22 @@ status_t GLConsumer::doGLFenceWaitLocked() const {
                 return err;
             }
         }
+    } else if (SyncFeatures::getInstance().useFenceSync()) {
+        EGLSyncKHR fence = eglCreateSyncKHR(dpy, EGL_SYNC_FENCE_KHR, NULL);
+        if (fence == EGL_NO_SYNC_KHR) {
+            ST_LOGE("doGLFenceWait: error creating fence: %#x",
+                    eglGetError());
+            return UNKNOWN_ERROR;
+        }
+        EGLint result = eglClientWaitSyncKHR(dpy, fence, 0, 1000000000);
+        if (result == EGL_FALSE) {
+            ST_LOGE("doGLFenceWait: error waiting for fence: %#x", eglGetError());
+            return UNKNOWN_ERROR;
+        } else if (result == EGL_TIMEOUT_EXPIRED_KHR) {
+            ST_LOGE("sdoGLFenceWait: timeout waiting for fence");
+            return TIMED_OUT;
+        }
+        eglDestroySyncKHR(dpy, fence);
     }
 
     return NO_ERROR;
-- 
1.8.0

